Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm

```diff
@@ -198,33 +198,68 @@ static bool isCopyDisplayedPixelBufferAvailable()
     return SourceBufferParserWebM::isContentTypeSupported(parameters.type);
 }
 
-void MediaPlayerPrivateWebM::load(const String& url)
+void MediaPlayerPrivateWebM::setPreload(MediaPlayer::Preload preload)
 {
-    ALWAYS_LOG(LOGIDENTIFIER);
+    ALWAYS_LOG(LOGIDENTIFIER, " - ", static_cast<int>(preload));
+    m_preload = preload;
+    if (m_assetURL.isEmpty())
+        return;
 
     auto player = m_player.get();
     if (!player)
         return;
 
+    if (m_preload >= MediaPlayer::Preload::MetaData && !m_resourceClient) {
+        if (!createResourceClient(m_assetURL)) {
+            ERROR_LOG(LOGIDENTIFIER, "could not create resource client");
+            setNetworkState(MediaPlayer::NetworkState::NetworkError);
+            setReadyState(MediaPlayer::ReadyState::HaveNothing);
+        }
+    }
+
     auto mimeType = player->contentMIMEType();
     if (mimeType.isEmpty() || !mimeTypeCache().contains(mimeType)) {
         ERROR_LOG(LOGIDENTIFIER, "mime type = ", mimeType, " not supported");
         setNetworkState(MediaPlayer::NetworkState::FormatError);
         return;
     }
 
+    if (preload > MediaPlayer::Preload::MetaData) {
+        for (auto it = m_readyForMoreSamplesMap.begin(); it != m_readyForMoreSamplesMap.end(); ++it)
+            notifyClientWhenReadyForMoreSamples(it->first);
+    }
+}
+
+void MediaPlayerPrivateWebM::load(const String& url)
+{
+    ALWAYS_LOG(LOGIDENTIFIER);
+
+    setNetworkState(m_preload == MediaPlayer::Preload::None ? MediaPlayer::NetworkState::Idle : MediaPlayer::NetworkState::Loading);
+    setReadyState(MediaPlayer::ReadyState::HaveNothing);
+
+    m_assetURL = URL({ }, url);
+
+    setPreload(m_preload);
+}
+
+bool MediaPlayerPrivateWebM::createResourceClient(const URL& url)
+{
+    RefPtr player = m_player.get();
+    if (!player)
+        return false;
+
     ResourceRequest request(url);
     request.setAllowCookies(true);
```

Comment:
This is existing code but it feels a bit weird that this is handled there and not for container agnostic code.

Reply:
+1

Reply:
I'm not sure I understand the comment here. 
m_player is a WeakPtr, and we need to access it to create the ResourceLoader.

Reply:
the weakptr to the player is owned by each MediaPlayerPrivate ; not checking there and have it be owned at the higher level would be a big change, including having to typically downcast in some circumstances.

Reply:
It is the call to setAllowCookies that feels a bit out of place, this could be best addressed in a follow-up.